### ISSUE TITLE
InternalPort when trying to publish on MQTT

### ISSUE THREAD

**Issue by kehusa** [HEADS-project/forum#14]
Hi,

I am trying to publish on to MQTTChan, but nothing happens. When debugging i get to:
 onSend: function (fromPortPath, destPortPaths, msg, callback) {
        var model = this.getKevoreeCore().getDeployModel() || this.getKevoreeCore().getCurrentModel();

```
    destPortPaths.forEach(function (path) {
        var targetNode = model.findByPath(path).eContainer().eContainer().name;
        if (targetNode === this.getNodeName()) {
            // local message
            this.localDispatch(msg);
        } else {
            // remote message
            if (this.client) {
                this.client.publish(this.uuid+path, msg+'');
            }
        }
    }.bind(this));
}
```

});

where destPortPaths is empty. Nothing happens. Seems like it gets called from:
 internalSend: function (outputPath, msg, callback), but shouldn't this be an external send?

**Comment by maxleiko** [comment]
Can you give me the model to reproduce this ?

**Comment by kehusa** [comment]
Sent you a zip file by email

**Comment by maxleiko** [comment]
@kehusa yep, your model seems wrong.
You have connected your output port to a MQTTChannel, and this channel is not connected to any input ports. That's why `destPortPaths` is empty.. no input connected. I guess ?

**Comment by kehusa** [comment]
OK, I understand, but I had hoped to be able to use the MQTT channel to publish messages out of the HEADS system. I do not know if there are any receivers of this info, and I do not care either. If i need to connect the channel to something, it should have been kind of a virtual node that doesn't need to exist.

**Comment by maxleiko** [comment]
Oh, that's true. I will add a dictionary attribute to the MQTTChannel something like `sendAnyway` to handle this use case

**Comment by kehusa** [comment]
Great :-)

**Comment by brice-morin** [comment]
Yes, we actually discussed that capability of being able to communicate with non-HEADS systems via MQTT and maybe HTTP/REST. I actually sent an e-mail to @maxleiko, @barais and @ffleurey about that on 2014-08-22 following a meeting SINTEF had with TellU. Good to see it will be supported, as we have to be careful not to impose a HEADS vendor-lockin on our systems, and instead by open.

**Comment by kehusa** [comment]
One more observation, when I connected the input part of the port to the same mqtt channel like this:
bind node0.HomeGW_0.HomeGW_server_smarttracker_out chanmqtt
bind node0.HomeGW_0.HomeGW_server_smarttracker_in chanmqtt

which I think is reasonable since the mqtt channel is bidirectional (pub and sub) the messages was just directed to the same node without going out on the mqtt broker. I think the messages always should be published on the broker. Btw, this would perhaps solve the previous problem, but it requires that the port is defined to receive some messages in thingML. Otherwise it will not be defined in the model.

**Comment by maxleiko** [comment]
Yes that was by choice (to save time because of network latency) but I can also change that if it is needed.
https://github.com/kevoree/kevoree-js-chan-mqtt/blob/master/lib/MQTTChannel.js#L112

**Comment by kehusa** [comment]
I think that by using the MQTTChannel you expect messages always being published. You cannot know who else is subscribing for what you publish. Using the topic wildcards, other can subscribe for stuff that you publish not knowing your uuid.

**Comment by maxleiko** [comment]
@kehusa Okay so, finally I won't modify the **MQTTChannel**, instead I've created 2 new component types:
- **MQTTSubClient**: proxy received messages from the MQTT broker it is connected (using the dictionary attributes) to its output port "onMsg"
- **MQTTPubClient**: publish received messages from its input port "pub" to the MQTT broker it is connected (using dictionary attributes)

**Comment by kehusa** [comment]
I have successfully used the PubClient, but I have encountered a problem
related to ThingML and Kevoree integration.
When I send messages through ThingML ports into a pure kevoree component,
the messages has the format: {"message":"reading","port":"smarttracker_c",
"value":122}. This is what the MQTTPubClient in turn publish on the MQTT
bus. However, since I am "leaving" HEADS technology, I would like the
ThingML "stuff" to be removed so that only the value part is what is being
published. It might be the LocalChannel (or HEADSChannel)'s responsibility
to do this. What do you think?

2015-04-28 12:58 GMT+02:00 Maxime Tricoire notifications@github.com:

> @kehusa https://github.com/kehusa Okay so, finally I won't modify the
> _MQTTChannel_, instead I've created 2 new component types:
> - _MQTTSubClient_: proxy received messages from the MQTT broker it is
>   connected (using the dictionary attributes) to its output port "onMsg"
> - _MQTTPubClient_: publish received messages from its input port "pub"
>   to the MQTT broker it is connected (using dictionary attributes)
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HEADS-project/forum/issues/14#issuecomment-97015671.

## 

| Knut Eilif Husa | CTO | Tellu AS | Hagaløkkveien 13, 1383 Asker - NORWAY|
Direct: +47 452 49496 | www.tellu.no |

This message contains confidential information and is intended only for the
individual named. If you are not the named addressee you should not
disseminate, distribute or copy this e-mail. Please notify the sender
immediately by e-mail if you have received this e-mail by mistake and
delete this e-mail from your system.

**Comment by maxleiko** [comment]
Get rid of the "protocol-overhead" manually in your component just before sending it, or proxy the message into another component that will do this job for you.

``` json
{ "message" : "reading", "port" : "smarttracker_c", "value" : 122 }
```

This is only useful for ThingML ? Well, added by ThingML, but pointless for you in the end, is that it ?

@brice-morin any idea ?

**Comment by brice-morin** [comment]
Well, we never know how the external API will look like. You only need the value, some other external APIs might expect XML or JSON with different formats, etc. We could either make a special compiler for the TELLU case, that generates what do you expect (driven by some annotation) or you could implement a proxy component, either on the ThingML side or in plain Kevoree. The latter solution is the most generic and it makes sense to have a proxy complement explicit to document the fact we use a remote server.

**Comment by brice-morin** [comment]
Also it depends on how many proxy you need. If just a few, it can be done manually. If a lot, then the generative approach makes more sense (less errors once the compiler is done right).

**Comment by sdalgard** [comment]
When using ThingML for the SINTEFboard we need to control the serialization used for that target. 
I assumed that this will be fixed by using port attributes specifying coding for that port. 
The target has to have support for the specific coding specified.

In the WP2 presentation from SINTEF meeting in April the ambition is described:
![image](https://cloud.githubusercontent.com/assets/2270169/7696700/ef2a5156-fdff-11e4-9075-721e61bed188.png)

**Comment by brice-morin** [comment]
Yes, I need a few concrete examples of serialized messages to help you (platform expert) write the proper compiler fragments.

**Comment by kehusa** [comment]
For the Tellu case I also think that an annotation to the port would be the
best way to go. The proxy solution would require me to write a ThingML
component that still needs to call out to a kevoree component. Actually I
don't see how I from within a ThingML component can reach out to the
Kevoree ports. I guess I shouldn't either since the Kevoree component is a
container for ThingML.

Knut Eilif

2015-05-19 9:22 GMT+02:00 Brice Morin notifications@github.com:

> Yes, I need a few concrete examples of serialized messages to help you
> (platform expert) write the proper compiler fragments.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HEADS-project/forum/issues/14#issuecomment-103378119.

## 

| Knut Eilif Husa | CTO | Tellu AS | Hagaløkkveien 13, 1383 Asker - NORWAY|
Direct: +47 452 49496 | www.tellu.no |

This message contains confidential information and is intended only for the
individual named. If you are not the named addressee you should not
disseminate, distribute or copy this e-mail. Please notify the sender
immediately by e-mail if you have received this e-mail by mistake and
delete this e-mail from your system.

**Comment by kehusa** [comment]
So how do we proceed to make this "external/remove protocol" annotation to
the port?

Knut Eilif

2015-05-19 9:35 GMT+02:00 Knut Eilif Husa knut.eilif.husa@tellu.no:

> For the Tellu case I also think that an annotation to the port would be
> the best way to go. The proxy solution would require me to write a ThingML
> component that still needs to call out to a kevoree component. Actually I
> don't see how I from within a ThingML component can reach out to the
> Kevoree ports. I guess I shouldn't either since the Kevoree component is a
> container for ThingML.
> 
> Knut Eilif
> 
> 2015-05-19 9:22 GMT+02:00 Brice Morin notifications@github.com:
> 
> > Yes, I need a few concrete examples of serialized messages to help you
> > (platform expert) write the proper compiler fragments.
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/HEADS-project/forum/issues/14#issuecomment-103378119
> > .
> 
> ## 
> 
> | Knut Eilif Husa | CTO | Tellu AS | Hagaløkkveien 13, 1383 Asker -
> NORWAY| Direct: +47 452 49496 | www.tellu.no |
> 
> This message contains confidential information and is intended only for
> the individual named. If you are not the named addressee you should not
> disseminate, distribute or copy this e-mail. Please notify the sender
> immediately by e-mail if you have received this e-mail by mistake and
> delete this e-mail from your system.

## 

| Knut Eilif Husa | CTO | Tellu AS | Hagaløkkveien 13, 1383 Asker - NORWAY|
Direct: +47 452 49496 | www.tellu.no |

This message contains confidential information and is intended only for the
individual named. If you are not the named addressee you should not
disseminate, distribute or copy this e-mail. Please notify the sender
immediately by e-mail if you have received this e-mail by mistake and
delete this e-mail from your system.

**Comment by brice-morin** [comment]
Well, it is always a bit tricky to generate code without knowing exactly what to generate... If possible, I suggest that you manually hack the currently generated code so that it fits your need and so that I have one example of what to generate. Put comments in every places you modified, and then send me your original ThingML project and the modified code. Then I can make a first version of the compiler + annotations on your ThingML models, which you can then take over. I guess the TellU case is in JS?

For the SINTEF board, I guess in C, @sdalgard should see with @ffleurey

**Comment by kehusa** [comment]
OK, I will do that. Yes, it is currently JS, but I think this compiler
directive should work for both JS and Java. It is very general. Just remove
the { "message" : "reading", "port" : "smarttracker_c", "value" : 122 }
from what is emitted from the port. I only want the 122 emitted from the
port.

Knut Eilif

2015-05-19 9:43 GMT+02:00 Brice Morin notifications@github.com:

> Well, it is always a bit tricky to generate code without knowing exactly
> what to generate... If possible, I suggest that you manually hack the
> currently generated code so that it fits your need and so that I have one
> example of what to generate. Put comments in every places you modified, and
> then send me your original ThingML project and the modified code. Then I
> can make a first version of the compiler + annotations on your ThingML
> models, which you can then take over. I guess the TellU case is in JS?
> 
> For the SINTEF board, I guess in C, @sdalgard
> https://github.com/sdalgard should see with @ffleurey
> https://github.com/ffleurey
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HEADS-project/forum/issues/14#issuecomment-103383142.

## 

| Knut Eilif Husa | CTO | Tellu AS | Hagaløkkveien 13, 1383 Asker - NORWAY|
Direct: +47 452 49496 | www.tellu.no |

This message contains confidential information and is intended only for the
individual named. If you are not the named addressee you should not
disseminate, distribute or copy this e-mail. Please notify the sender
immediately by e-mail if you have received this e-mail by mistake and
delete this e-mail from your system.

**Comment by brice-morin** [comment]
OK, perfect. Yes, the annotations should apply to all target languages. But the way we handle it concretely for the different targets might vary

**Comment by sdalgard** [comment]
In addition to the above mentioned plan it would be nice to define a port as transparent octet channel.
In this way ThingML code can be used to program a new protocol on a port. The port will only support "byte in" or "byte out" to increase efficiency array of byte can also be supported.

**Comment by brice-morin** [comment]
I am starting implementing a solution to support the TellU case. I guess I can fairly easily make something that works for that case, but as for the generality of the approach, it might be some issues in the case you define a ThingML port like this:

```
message temp(t : Integer)
message light(l : Integer)
required port weather 
@external "..."
{
    sends temp, light
}
```

Obviously, if I just pass the value to Kevoree, we cannot know whether it is a light or temperature measurement.

Also, it might be useful in some cases that this port is both connected to an external server and to an internal (HEADS-managed) component. If we remove the HEADS-specific serialization (used internally to communicate among HEADS-component), then we cannot support that use case (as we need information about the ThingML message and port to be able to internally route the message).

So I think I will go for a generative approach that generates proxy. Basically, for each `@external` port, I will generate one additional Kevoree component. On one side it will expose "HEADS-specific" ports (which send or receive message with "internal" serialization) and on the other side it will provide "neutral" ports (in a 1 ThingML message - 1 Kevoree port fashion) that just send the values. I'll make a picture describing the approach, as I suspect my explanation here are not too clear :-)

**Comment by kehusa** [comment]
Ok, I understand. This is probably a smart way to fix this issue. I think
the @external annotation is well-descriptive for what the port does to the
messages.

Knut Eilif

2015-05-20 10:00 GMT+02:00 Brice Morin notifications@github.com:

> I am starting implementing a solution to support the TellU case. I guess I
> can fairly easily make something that works for that case, but as for the
> generality of the approach, it might be some issues in the case you define
> a ThingML port like this:
> 
> message temp(t : Integer)
> message light(l : Integer)
> required port weather
> @external "..."
> {
>     sends temp, light
> }
> 
> Obviously, if I just pass the value to Kevoree, we cannot know whether it
> is a light or temperature measurement.
> 
> Also, it might be useful in some cases that this port is both connected to
> an external server and to an internal (HEADS-managed) component. If we
> remove the HEADS-specific serialization (used internally to communicate
> among HEADS-component), then we cannot support that use case (as we need
> information about the ThingML message and port to be able to internally
> route the message).
> 
> So I think I will go for a generative approach that generates proxy.
> Basically, for each @external port, I will generate one additional
> Kevoree component. On one side it will expose "HEADS-specific" ports (which
> send or receive message with "internal" serialization) and on the other
> side it will provide "neutral" ports (in a 1 ThingML message - 1 Kevoree
> port fashion) that just send the values. I'll make a picture describing the
> approach, as I suspect my explanation here are not too clear :-)
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HEADS-project/forum/issues/14#issuecomment-103801019.

## 

| Knut Eilif Husa | CTO | Tellu AS | Hagaløkkveien 13, 1383 Asker - NORWAY|
Direct: +47 452 49496 | www.tellu.no |

This message contains confidential information and is intended only for the
individual named. If you are not the named addressee you should not
disseminate, distribute or copy this e-mail. Please notify the sender
immediately by e-mail if you have received this e-mail by mistake and
delete this e-mail from your system.

**Comment by brice-morin** [comment]
OK, I made a first proof-of-concept. Basically if you have

```
message temp(t : Integer)
message light(l : Integer)
message both(t : Integer, l : Integer)
required port weather 
@external "..."
{
    sends temp, light, both
}
```

It will generate the normal ThingML/KevoreeJS component + a proxy that looks like:

```
...
in_weather: function (msg) {
  var json = JSON.parse(msg);
  if (json.message === temp) {
    this.out_weather_weather_temp(json.temp);
  }else if (json.message === light) {
    this.out_weather_weather_light(json.light);
  }else if (json.message === both) {
    this.out_weather_weather_both(json.temp + "; " + json.light);
}},
out_weather_weather_temp: function(msg) {/* This will be overwritten @runtime by Kevoree JS */},
out_weather_weather_light: function(msg) {/* This will be overwritten @runtime by Kevoree JS */},
out_weather_weather_both: function(msg) {/* This will be overwritten @runtime by Kevoree JS */}}});
...
```

It will also add some lines in the generated KevScript to connect the proxy:

```
//instantiate Kevoree/ThingML proxy components
add node0.weather_0 : my.package.timer
add chan0 : LocalChannel
bind node0.weather.out_weather_weather_weather_out chan0
bind node0.weather_0.in_weather chan0
```

> Currently the proxy is generated in the same project, I need to generate it in a separate project

**Comment by maxleiko** [comment]
You must wrap your call to `JSON.parse` in a try/catch  
Otherwise the whole thing might blow up

**Comment by brice-morin** [comment]
well, the JSON is generated by myself, hence correct ;-)

**Comment by maxleiko** [comment]
Yes but because it is a Kevoree component, anyone can plug something to the `weather` input port and make it explode :)

**Comment by brice-morin** [comment]
hmmmmm.... I see... I will fix that + the other bug we discuss on the chat

**Comment by brice-morin** [comment]
OK, I updated JS compiler and KevoreeJS compiler. JS compiler does not rely on the internal JSON serialization anymore, making it simpler to wrap into KevoreeJS. Kevoree JS should now by default only send values on the ports. Note that I went for a 1-thingml-message<->1-kevoree-port mapping to avoid having to serialize message name into the message. Just the value is serialized. In case the message holds several values, it will be serialized as: `value1; value2`
